rcl/test/rcl/test_events.cpp

```diff
@@ -0,0 +1,546 @@
+    ASSERT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+    // init subscription event
+    subscription_event = rcl_get_zero_initialized_event();
+    ret = rcl_subscription_event_init(&subscription_event, &subscription, sub_event_type);
+    ASSERT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+    // wait for discovery
+    // total wait time of 10 seconds, if never ready
+    size_t max_iterations = 100;
+    milliseconds wait_period(10);
+    size_t iteration = 0;
+    do {
+      iteration++;
+      size_t count = 0;
+      rcl_ret_t ret = rcl_subscription_get_publisher_count(&subscription, &count);
+      ASSERT_EQ(ret, RCL_RET_OK);
+      if (count > 0) {
+        break;
+      }
+      std::this_thread::sleep_for(wait_period);
+    } while (iteration < max_iterations);
+  }
+
+  void tear_down_publisher_subscriber()
+  {
+    rcl_ret_t ret;
+
+    ret = rcl_event_fini(&subscription_event);
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+    ret = rcl_subscription_fini(&subscription, this->node_ptr);
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+    ret = rcl_event_fini(&publisher_event);
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+    ret = rcl_publisher_fini(&publisher, this->node_ptr);
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  }
+
+  void TearDown()
+  {
+    rcl_ret_t ret;
+
+    ret = rcl_node_fini(this->node_ptr);
+    delete this->node_ptr;
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+    ret = rcl_shutdown(this->context_ptr);
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+    ret = rcl_context_fini(this->context_ptr);
+    delete this->context_ptr;
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  }
+
+protected:
+  rcl_context_t * context_ptr;
+  rcl_node_t * node_ptr;
+
+  rcl_publisher_t publisher;
+  rcl_event_t publisher_event;
+  rcl_subscription_t subscription;
+  rcl_event_t subscription_event;
+  bool is_unsupported, is_opensplice, is_fastrtps;
+  const char * topic = "rcl_test_publisher_subscription_events";
+  const rosidl_message_type_support_t * ts;
+};
+
+rcl_ret_t
+wait_for_msgs_and_events(
+  rcl_subscription_t * subscription,
+  rcl_event_t * subscription_event,
+  rcl_event_t * publisher_event,
+  rcl_context_t * context,
+  int64_t period_ms,
+  bool & msg_ready,
+  bool & subscription_event_ready,
+  bool & publisher_event_ready)
+{
+  msg_ready = false;
+  subscription_event_ready = false;
+  publisher_event_ready = false;
+
+  int num_subscriptions = (nullptr == subscription ? 0 : 1);
+  int num_events = (nullptr == subscription_event ? 0 : 1) + (nullptr == publisher_event ? 0 : 1);
+
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t ret = rcl_wait_set_init(&wait_set, num_subscriptions, 0, 0, 0, 0, num_events,
+      context, rcl_get_default_allocator());
```

Comment:
nitpick: please do not indent more than one level (in this case two spaces)

Reply:
Fixed.

Reply:
Done

Reply:
Seems like this got added back in the latest commit?

Reply:
Fixed.